File: 1049 Last Stone Weight II/1049 Last Stone Weight II.py
"""
You are given an array of integers stones where stones[i] is the weight of the ith stone.
We are playing a game with the stones. On each turn, we choose any two stones and smash them together. Suppose the stones have weights x and y with x <= y. The result of this smash is:

    If x == y, both stones are destroyed, and
    If x != y, the stone of weight x is destroyed, and the stone of weight y has new weight y - x.

At the end of the game, there is at most one stone left.

Return the smallest possible weight of the left stone. If there are no stones left, return 0.

Example 1:

Input: stones = [2,7,4,1,8,1]
Output: 1
Explanation:
We can combine 2 and 4 to get 2, so the array converts to [2,7,1,8,1] then,
we can combine 7 and 8 to get 1, so the array converts to [2,1,1,1] then,
we can combine 2 and 1 to get 1, so the array converts to [1,1,1] then,
we can combine 1 and 1 to get 0, so the array converts to [1], then that's the optimal value.

Example 2:

Input: stones = [31,26,33,21,40]
Output: 5
"""
class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        l = stones
        half_sum = sum(l) // 2
        sums = {0}
        for process in l:
            new_sums = set()
            for num in sums:
                if num + process <= half_sum:
                    new_sums.add(num + process)
            sums |= new_sums
        return sum(l) - 2 * max(sums)

File: 1049 Last Stone Weight II/1049 Last Stone Weight II.go
func lastStoneWeightII(stones []int) int {
	total := 0
	for _, v := range stones {
		total += v
	}
	half := total >> 1
	dp := make([]int, half+1)
	sort.Ints(stones)
	for _, v := range stones {
		for i := half; i >= v; i-- {
			dp[i] = max(dp[i], dp[i-v] + v)
		}
	}
	return total - 2*dp[half]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
